### PR TITLE
Adding the "Readonly references" proposal

### PR DESCRIPTION
**Pull request by VSadov**

### CODE REVIEW

**Comment by VSadov** [comment]
@gafter @jaredpar @MadsTorgersen @OmarTawfik  - please take a look.

File: proposals/readonly-ref.md [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+            v1 = default(Vector3);
+
+            // not OK!!
+            v1.X = 0;
+
+            // not OK!!
+            foo(ref v1.X);
+
+            // OK
+            return new Vector3(v1.X +v2.X, v1.Y + v2.Y, v1.Z + v2.Z);
```
Comment (OmarTawfik) [inline proposals/readonly-ref.md]: I'd add another example about passing to a second "readonly ref" method.

File: proposals/readonly-ref.md [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+`in` would not be allowed in combination with `out` or with anything that `out` does not combine with.
+For the purpose of OHI (Overloading, Hiding, Implementing), `in` will behave similarly to an `out` parameter. All the same rules apply, in particular it is not permitted to overload on `ref/out/in` differences.
+
+For the purpose of binding and overload resolution `in` behaves similarly to `out` as well. `in` is basically just a new element of `RefKind` enum, in addition to current `{None, Ref, Out}`.
+
+For the purpose of generic variance, `in` is nonvariant.
+
+## Use of `in` at call sites.
+
+Unlike `out` parameters, `in` does not need to be matched with an LValue(*) argument. At the call-site `in` parameters are no different from regular by-value parameters.
```
Comment (OmarTawfik) [inline proposals/readonly-ref.md]: I'd rather match ref and out by explicitly stating that in the call site.

File: proposals/readonly-ref.md [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+```C#
+        struct ImmutableArray<T>
+        {
+            private readonly T[] array;
+
+            public readonly ref T RefAt(int i)
+            {
+                // returning a readonly ref 
+                return ref this.r1;
```
Comment (OmarTawfik) [inline proposals/readonly-ref.md]: Same here. I suggest "return readonly ref this.r1" syntax. A bit lengthy but IMO much more clearer to the reader, and more consistent with "ref" and "out" features.

**Comment by jaredpar** [comment]
CC @dotnet/csharplangdesign

**Pull request merged**